sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySchemaTransformReadConfiguration.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.beam.sdk.schemas.annotations.DefaultSchema;
+import org.apache.beam.sdk.schemas.io.InvalidConfigurationException;
+
+/**
+ * Configuration for reading from BigQuery.
+ *
+ * <p>This class is meant to be used with {@link BigQuerySchemaTransformReadProvider}.
+ *
+ * <p><b>Internal only:</b> This class is actively being worked on, and it will likely change. We
+ * provide no backwards compatibility guarantees, and it should not be implemented outside the Beam
+ * repository.
+ */
+@DefaultSchema(AutoValueSchema.class)
+@AutoValue
+public abstract class BigQuerySchemaTransformReadConfiguration {
+
+  private static final boolean DEFAULT_USE_STANDARD_SQL = true;
+
+  /**
+   * Instantiates a {@link BigQuerySchemaTransformReadConfiguration.Builder} from the SQL query.
+   *
+   * <p>The configuration defaults to useStandardSql=true.
+   */
+  public static Builder createQueryBuilder(String query) {
+    return defaultBuilder().setQuery(query).setJobType(JobType.QUERY);
+  }
+
+  /**
+   * Instantiates a {@link BigQuerySchemaTransformReadConfiguration.Builder} to support BigQuery
+   * extract jobs. See the getTableSpec() getter for details.
+   */
+  public static Builder createExtractBuilder(String tableSpec) {
+    return defaultBuilder().setTableSpec(tableSpec).setJobType(JobType.EXTRACT);
+  }
+
+  /**
+   * Instantiates a {@link BigQuerySchemaTransformReadConfiguration.Builder} to support BigQuery
+   * extract jobs.
+   */
+  public static Builder createExtractBuilder(TableReference tableSpec) {
+    if (tableSpec.getProjectId().isEmpty()) {
+      return createExtractBuilder(
+          String.format("%s.%s", tableSpec.getDatasetId(), tableSpec.getTableId()));
+    }
+    return createExtractBuilder(
+        String.format(
+            "%s:%s.%s",
+            tableSpec.getProjectId(), tableSpec.getDatasetId(), tableSpec.getTableId()));
+  }
+
+  private static Builder defaultBuilder() {
+    return new AutoValue_BigQuerySchemaTransformReadConfiguration.Builder()
+        .setJobType(JobType.UNSPECIFIED)
+        .setQuery("")
+        .setQueryLocation("")
+        .setTableSpec("")
+        .setUseStandardSql(DEFAULT_USE_STANDARD_SQL);
+  }
+
+  /** Configures the BigQuery job type. */
+  abstract JobType getJobType();
+
+  /** Configures the BigQuery read job with the SQL query. */
+  public abstract String getQuery();
+
+  /**
+   * Specifies a table for a BigQuery read job as "[project_id]:[dataset_id].[table_id]" or
+   * "[dataset_id].[table_id]" for tables within the current project.
+   */
+  public abstract String getTableSpec();
+
+  /** BigQuery geographic location where the query job will be executed. */
+  public abstract String getQueryLocation();
+
+  /** Enables BigQuery's Standard SQL dialect when reading from a query. */
+  public abstract Boolean getUseStandardSql();
+
+  /** Instantiates a {@link BigQueryIO.TypedRead} from the configuration. */
+  public BigQueryIO.TypedRead<TableRow> toTypedRead() {
```

Comment:
Would it make sense to put this logic all into the SchemaTransform and then just make this class a configuration class? I think having some of the logic split between them is a bit odd.